onboarding: Add a one-time intro tooltip for the "go to conversation" button.

saubhagya-patel opened the pull request:
This PR adds a one-time intro tooltip for the "go to conversation" button. It includes the following changes:
 - Add a one-time go-to conversation tooltip as an onboarding step.
 - Implement tests to verify the tooltip.
 
 Fixes #32166.
[CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/.22go.20to.20conversation.22.20button.20intro/near/1969306)
 
 The tooltip is displayed highlighting the "go to conversation" button without being hovered.
 Conditions for showing the tooltip as mentioned in the issue #32166:

- The "Go to conversation" button is active.
- You just confirmed a topic recipient change to a topic that we believe contains messages and is not the current view.
- You have no compose banners showing.
- You haven't clicked the button (which is the mechanism that would dismiss it).
- You haven't been shown the same tooltip before.


<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

saubhagya-patel commented:
Tagging @userAdityaa for a review following [this]((https://github.com/zulip/zulip/pull/32169#issuecomment-2552974582)) comment.

saubhagya-patel commented:
Following the questions mentioned in the [issue](https://github.com/zulip/zulip/issues/32166):
- Do we want to show this tooltip twice (once for DMs and once for topics), as specced? Or just once?
   - Using this implementation the tooltip will only be shown once similar to the compose banners being shown in interleaved views.

- Should this tooltip be shown to existing users?
    - Please correct me if I am wrong, I think the tooltip will show once to the existing users as well with the updated set of changes until the Zulip server has disabled the tutorial.

saubhagya-patel commented:
Tagging @userAdityaa for a review following [this](https://github.com/zulip/zulip/pull/32169#issuecomment-2552974582) comment.

alya commented:
Playing with it, I like showing it just once.

alya commented:
We can decide a bit later whether to mark existing users as having already completed this onboarding step.

alya commented:
@prakhar1144 Are you up for reviewing this one?

alya commented:
This PR, rather than the issue, is the right place to post this comment:

> Following the [continuing unfinished work documentation](https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html), please note that PR https://github.com/zulip/zulip/pull/32773 takes a different approach than the one in this PR. Specifically, this PR adds the tooltip when updating the class of the "go-to conversation" button, whereas PR https://github.com/zulip/zulip/pull/32773 introduces the tooltip when the #compose-textarea is focused.

pratik-pc commented on web/src/compose_actions.ts:
```diff
@@ -455,6 +459,7 @@ export let cancel = (): void => {
     compose_state.set_message_type(undefined);
     compose_pm_pill.clear();
     $(document).trigger("compose_canceled.zulip");
+    compose_tooltips.hide_go_to_conversation_button_intro_tooltip();
```
Here, we are calling it every time as well.

saubhagya-patel replied:
This is now fixed. Thank you.

saubhagya-patel commented on web/src/compose_setup.js:
```diff
@@ -490,6 +491,16 @@ export function initialize() {
+            onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY.has(
+                "intro_go_to_conversation_button_tooltip",
+            )
+        ) {
```
That sounds like a reasonable approach to me and I have fixed it. Thank you

pratik-pc commented on web/src/compose_tooltips.ts:
```diff
@@ -289,3 +292,70 @@ export function hide_compose_control_button_tooltips($row: JQuery): void {
+    // The "Go to conversation" button is active.
+    // There are no compose banners currently visible.
+    // The user hasn't clicked the button (which dismisses the tooltip).
+
+    // This helps in updating tooltip when switching between topics.
+    hide_go_to_conversation_button_intro_tooltip();
+
+    if (
+        !onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY.has("intro_go_to_conversation_button_tooltip")
+    ) {
+        return false;
+    }
```
I think we should call `hide_go_to_conversation_button_intro_tooltip` after the `onboarding_step` check here as well, unless there's a specific case you are trying to avoid.

saubhagya-patel replied:
I must have missed it, I apologize for the inconvenience caused. It is fixed in the subsequent push.

saubhagya-patel commented on web/src/compose_tooltips.ts:
```diff
@@ -289,3 +292,70 @@ export function hide_compose_control_button_tooltips($row: JQuery): void {
+    const faded_messages_exist = $(".focused-message-list .recipient_row").hasClass("message-fade");
+    if (!faded_messages_exist) {
+        return false;
+    }
```
Hmm, that sounds right, I will fix it in the subsequent push. Thank you.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- As mentioned in [this](https://github.com/zulip/zulip/pull/32773#discussion_r1894486694) comment, in the `compose_tooltips.can_show_go_to_conversation_button_intro_tooltip()` function, the `hide_go_to_conversation_button_intro_tooltip()` function is now called after checking if `onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY` has the go-to conversation button tooltip (`intro_go_to_conversation_button_tooltip`).

- As mentioned in [this](https://github.com/zulip/zulip/pull/32773#discussion_r1894481595) comment, to check whether a user is composing to the same topic/DM as the current view, I have updated the approach from using `faded_messages_exist` to confirming whether there is the same topic/DM in the `narrow_state` and compose box (similar to the function `compose_recipient.update_narrow_to_recipient_visibility()`).

- **Visual Changes:** Now, the tooltip is also shown in different views. Please check the attached GIF for the Recent conversations view.
  <details>
  <summary>GIF of tooltip in Recent conversations view.</summary>
   
   ![tooltip in recent conv](https://github.com/user-attachments/assets/ce0eba3e-e763-4883-aad8-dca622422a7f)

  </details>
  
@pratik-pc, I have made the required changes. Please let me know if any further adjustments are needed. Thank you.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- The function `can_show_go_to_conversation_button_intro_tooltip()` in the `compose_tooltips.ts` file is refactored following the [comment[1]](https://github.com/zulip/zulip/pull/32773#discussion_r1896099632) and [comment[2]](https://github.com/zulip/zulip/pull/32773#discussion_r1896099738).

- Following [this](https://github.com/zulip/zulip/pull/32773#discussion_r1896099713) comment, a separate [commit](https://github.com/zulip/zulip/pull/32773/commits/df179b0365622272754232a2147a3e74a55d1922) is added in which the logic of function `update_narrow_to_recipient_visibility()` is refactored into a separate function named `should_display_narrow_to_recipient()` in the `compose_recipient.ts` file.

- No visual changes are made.

@pratik-pc, thank you for the ongoing feedback. I hope the updated changes look good to you. Please let me know if any further adjustments are needed. Thank you.

saubhagya-patel commented on web/src/compose_tooltips.ts:
```diff
@@ -289,3 +292,91 @@ export function hide_compose_control_button_tooltips($row: JQuery): void {
+        ) {
+            return false;
+        }
+    }
```
Sure, that sounds better, I have implemented this in the following push.

pratik-pc commented:
You'd probably want to call `hide_go_to_conversation_button_intro_tooltip` or `show_go_to_conversation_button_intro_tooltip` when `narrow_to_compose_recipients` class is disabled and stream dropdown is changed to avoid getting in a bad ui state. Maybe test it out again for any ui issues.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- Following [this](https://github.com/zulip/zulip/pull/32773#issuecomment-2578726053) comment, I have updated the function calls to `hide_go_to_conversation_button_intro_tooltip()` and `show_go_to_conversation_button_intro_tooltip()`.

  - The function calls are updated to cover the UI bugs discovered in the previous implementation.
  
  - There was, however, a bug in the previous implementation where if we switched the channel recipient through the dropdown, the tooltip was not updated as the functions were called when the `textarea` was focused.
     <details>
       <summary>Screenshot of the bug.</summary>
    
    
       I found this bug when if we compose a message to the topic in the current view -> change to a different channel through dropdown -> tooltip appears -> again change the channel to the current view -> tooltip doesn't disappear.
    
       ![Screenshot 2025-01-16 211426](https://github.com/user-attachments/assets/acd48d39-b312-4203-8f89-295aa1347d8a)

     </details>



I tried my best to find and resolve any discoverable UI bugs in this implementation.
@pratik-pc, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required. Thank you.

pratik-pc commented:
![Screenshot from 2025-01-21 01-26-12](https://github.com/user-attachments/assets/76c569bd-bfaf-4125-91fa-8ac7977a325d)
In combined feed, there is a tooltip over compose banner.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- As stated in [this](https://github.com/zulip/zulip/pull/32773#issuecomment-2603894283) comment, the tooltip is no longer shown over a banner. I apologize for the inconvenience caused.

- A bug was discovered in the previous implementation: when a user clicked on the "Go to Conversation" button, the `onboarding_steps.post_onboarding_step_as_read(...)` function was executed, even if the tooltip was not visible at the time. This could result in the tooltip not being shown to the user at all. The functions have been updated to prevent this issue.

- No visible changes are made.

@pratik-pc, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required. Thank you.

pratik-pc commented:
@saubhagya-patel The tooltip isn't displayed once anymore. I am seeing it every time on changing the recipient even after narrowing.

alya commented:
@saubhagya-patel Please try testing your changes thoroughly before the next round of review, and write out how you tested. The [code review guide](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) may be helpful.

timabbott commented:
This can probably still be integrated for 10.0 if it's updated in the next couple days.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- Now, the tooltip is only displayed once to a user once a user dismisses it (addressed [this](https://github.com/zulip/zulip/pull/32773#issuecomment-2635037875) comment).

- Following [this](https://github.com/zulip/zulip/pull/32773#issuecomment-2635263845) comment, to test out the changes. I tested out with different banners and UI interactions across the app. I found some bugs which I have addressed in the following push of this PR.

- @pratik-pc [said](https://github.com/zulip/zulip/pull/32773#discussion_r1893183501): 
   > I think it's better to call `hide_go_to_conversation_button_intro_tooltip()` only if `ONE_TIME_NOTICES_TO_DISPLAY` contains it, instead of calling it on every narrow.
   
   I believe we won't need to check the above condition since we return early in both the functions `maybe_show_go_to_conversation_button_intro_tooltip()` and `hide_go_to_conversation_button_intro_tooltip()` on checking the same condition. Please let me know if I am wrong here.


@pratik-pc, could you please review this PR with the updated changes made?
Please let me know if any further changes are required.
Thank you.

pratik-pc commented:
I guess it is better to destroy the tippy instance when we won't have to show it again anymore.

Otherwise looks good to me, you can go ahead and ask for next review stage.

saubhagya-patel commented:
@prakhar1144, could you please review this PR once you have a moment?
Thank you.

timabbott commented:
@amanagr can do the next review as our tooltips expert.

saubhagya-patel commented:
@amanagr, the `hideOnClick` property is used because of the following reasons:
- Issue [#32166](https://github.com/zulip/zulip/issues/32166) states that: "You haven't clicked the button (which is the mechanism that would dismiss it)." as one of the conditions to show the tooltip. So, I figured it would be better if we show the tooltip to the user until the go-to conversation button is not clicked.

- Also, we have to show it only once; the tooltip will be hidden (if it is active) if a user clicks somewhere else other than the go-to conversation button but will not be posted as read for an onboarding step, hence it will be shown again, once the user focuses on the `<textarea>` element in the compose-box and we need to show it.

I hope I was able to explain my reasons behind using the property.
However, we can hide it if the user clicks outside it, as we use in other places. Can you please tell me how to proceed?
Thank you.

amanagr commented:
I think you can use `hideOnClick` and use `onHide` function call to exclude any click you don't want the tooltip to hide and `onHidden` to run any function after the tooltip is hidden (for example marking the onboarding step as complete).

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32773#issuecomment-2715280515) comment, I have used `hideOnClick` in the tooltip. However, I have used `onDestroy()` instead of `onHide()`. This was done because in the combined feed, when we open the compose box by clicking on a message, the `interleaved_view_messages_fading` banner appeared after the tooltip hence, we need to hide the tooltip to avoid displaying above it. Some UI issues were also found with the resolved topics banner. So, I decided to use the `onDestroy()` method which marks the onboarding step as complete and have used the .disable() method to hide the tooltip when a banner is shown to avoid UI bugs.

- Visual changes: _None_.

@amanagr, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

amanagr commented on web/src/compose_banner.ts:
```diff
@@ -83,6 +86,13 @@ export function append_compose_banner_to_banner_list(
     if ($banner.hasClass("warning") && has_error()) {
         return false;
     }
+    if (
+        onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY.has("intro_go_to_conversation_button_tooltip")
+    ) {
+        const element: tippy.ReferenceElement = the($(".conversation-arrow"));
+        const instance = element._tippy;
+        instance?.disable();
+    }
```
I don't understand what you are trying to do here. Why `disable` the instance?

amanagr commented on web/src/compose_recipient.ts:
```diff
@@ -298,6 +301,49 @@ export function handle_middle_pane_transition(): void {
+    const tippy_instance =
+        element._tippy ??
+        tippy.default(element, {
+            trigger: "manual",
```
Let's add `showoncreate`  and use `content` attribute to set the content. Then you won't need to do `setContent` and `show`.

amanagr commented on web/src/compose_recipient.ts:
```diff
@@ -298,6 +301,49 @@ export function handle_middle_pane_transition(): void {
+        parse_html($("#compose_go_to_conversation_button_tootltip_template").html()),
+    );
+
+    tippy_instance.enable();
```
Why `enable`?

saubhagya-patel replied:
Answered [here](https://github.com/zulip/zulip/pull/32773#issuecomment-2765380801).

amanagr commented on web/src/compose_send_menu_popover.js:
```diff
@@ -111,6 +112,7 @@ export function open_send_later_menu() {
             // When shown, we should give the modal focus to correctly handle keyboard events.
             const $send_later_modal_overlay = $("#send_later_modal .modal__overlay");
             $send_later_modal_overlay.trigger("focus");
+            popovers.hide_all();
         },
```
tooltip should automatically hide on outside click, this is not required.

amanagr commented on web/src/compose_setup.js:
```diff
@@ -565,6 +566,17 @@ export function initialize() {
 
     $("#compose").on("click", ".narrow_to_compose_recipients", (e) => {
         e.preventDefault();
+        if (
+            onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY.has(
+                "intro_go_to_conversation_button_tooltip",
+            )
+        ) {
+            const element = the($(".conversation-arrow"));
+            const instance = element._tippy;
+            if (instance !== undefined) {
+                instance.destroy();
+            }
+        }
```
This is not required, use property `hideOnClick` if the tooltip is not hidden.

amanagr commented on web/src/compose_setup.js:
```diff
@@ -583,6 +595,7 @@ export function initialize() {
         } else {
             compose_notifications.maybe_show_one_time_interleaved_view_messages_fading_banner();
         }
+        compose_recipient.maybe_show_go_to_conversation_button_intro_tooltip();
```
Should this be called in `setTimeout` to avoid race with our code which modified the class of the go to conversation button?

amanagr commented on web/templates/tooltip_templates.hbs:
```diff
@@ -21,6 +21,10 @@
     {{t 'Reply to selected conversation' }}
     {{tooltip_hotkey_hints "R"}}
 </template>
+<template id="compose_go_to_conversation_button_tootltip_template">
```
let's add `onboarding` prefix here.

amanagr commented:
You can check whether to show tooltip after a delay to avoid completing with banners being displayed. You can rely on `onDestroy` method but that's not different from calling it `onHide` fundamentally once you have got the mechanism working properly.

Please spend some time properly testing your changes.

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32773#discussion_r2007147312) comment, `onboarding` prefix is added to the tooltip template's id.

- Following [this](https://github.com/zulip/zulip/pull/32773#discussion_r2007145201) comment, `maybe_show_go_to_conversation_button_intro_tooltip()` is called in `setTimeout`.

- Following [this](https://github.com/zulip/zulip/pull/32773#discussion_r2007143886) comment, `hideOnClick` property is used, and as stated [here](https://github.com/zulip/zulip/pull/32773#discussion_r2007143062), we don't need to hide the tooltip manually.

- Following [this](https://github.com/zulip/zulip/pull/32773#discussion_r2007138331) comment, `showOnCreate` and `content` attributes are added.

- With respect to the comments, [[1]](https://github.com/zulip/zulip/pull/32773#discussion_r2007138569) and [[2]](https://github.com/zulip/zulip/pull/32773#discussion_r2007135300), regarding enabling and disabling the tooltip, it was done to remove the tooltip when a banner was shown without any click.  
  For example:  
  - Sending a message in the Starred messages view by pressing the Enter key shows the `jump_to_sent_message_conversation` banner, and the tooltip displays on top of it, which is a bug.  
    <details>
    <summary>Screenshot</summary>
    
    ![image](https://github.com/user-attachments/assets/33bd5509-c43a-40ca-8896-4666e5f8986c)  
    
    </details>  

  - Sending a message to a resolved topic not in the current view initially doesn't show any banner (`non_interleaved_view_messages_fading` banner is marked as read), hence the tooltip is displayed when we start composing. However, when we write some characters in the compose box, the `topic_resolved` banner is shown.  
    <details>
    <summary>Screenshot</summary>
    
    ![image](https://github.com/user-attachments/assets/161925c1-183c-4776-a362-4236946ef509)  
    
    </details>  

  - However, disabling and enabling the tooltip breaks the normal "Go to conversation" tooltip, as stated [here](https://github.com/zulip/zulip/pull/32773#issuecomment-2742760373).

@amanagr, thanks for your continued review.  
Since there are some bugs in this PR that I am currently unable to resolve, I have converted it into a draft PR. I will revisit this once I find a reasonable solution, unless someone else takes it over in the meantime.
